dev-docs/hail-query/key-encoding.md

```diff
@@ -0,0 +1,167 @@
+- byte - simply add a byte to the result, first padding an incomplete byte if necessary
+- bit - add a bit to the result, possibly leaving an incomplete byte. We must know statically how many bits are used in the byte.
+- pad - add `0`s to pad the last incomplete byte. This is safe (prefix-free) because the number of used bits is a (statically known) constant. We use this to ensure the number of used bits is known statically.
+	
+Types:
+- missingness
+  - treat as a type constructor `optional<T>`, i.e. base types don't encode missingness. Emits a single bit in the encoding. Can invert this bit to control whether missing values come first or last in the ordering. If missing, nothing is emitted after.
+- sort-order
+  - treat reversing the default ordering as a type constructor `reverse<T>`
+  - simply inverts the encoding bitwise
+- primitive types
+  - same as in datafusion, encoding has same size as original type
+  - signed integers - flip the sign bit
+  - floating point numbers - if sign bit is set, invert all bits, otherwise only flip the sign bit
+- arrays
+  - before each element and after last element, emit continuation bit (0 if no more elements)
+  - pad before each element. This prevents a variable number of missing bits packing into a byte
+- strings and byte-arrays
+  - simply use null-terminated strings (being careful to do this in a unicode-safe way)
+- structs
+  - simply concatenate element encodings. safe because codes are prefix-free
+- key structs
+  - support variable length "interval endpoints"
+    - e.g. for a key type `struct<t1, t2>`, the interval `[{a}, {a, b})` contains all keys with first field `a` and second field less than `b`. We break it into two "interval endpoints", `({a}, -1)` and `({a, b}, -1)`, which consist of a struct value which is a prefix of the key struct type, and a "sign". In this case, both endpoints "lean left".
+    - needed for working with partitioners at runtime
+  - like an array with fixed but heterogenous types and a max length
+  - before each element and after last element, emit two continuation bits
+    - `00` - end of key, leans left (less than all longer keys with this prefix)
+    - `01` - continue, or after last key field of actual key value (not interval endpoint)
+      - unambiguous because key value can't terminate early, and can't continue past last key field (max length)
+    - `11` - end of key, leans right (greater than all longer keys with this prefix)
+  - after each element, pad
+
+# Implementation sketch
+Concretely, we define a function `emit(value: SValue, trailingBits: Int)` which appends the encoding of `value` to an output buffer using the interface `emitByte(b: Value[Byte])` (along with `emitInt`, etc), which can only be called with 0 trailing bits, `emitBit(b: Value[Boolean], trailingBits: Int)`, and `pad()`. Note that this interface forces the caller to track the statically known number of trailing bits.
```

Comment:
I think it would really help if you included a mock-up of emitBit. I don't really follow what `trailingBits` is supposed to do